File: shunting_yard.py
import sys
import enum
import dataclasses


class ParseError(Exception):
    pass

# TokenType is not an enum anymore. Because we need a precedence for every operator.
# We can utilize zero precedence to distinguish operators from other token types.
@dataclasses.dataclass()
class TokenType:
    precedence: int

UMINUS = TokenType(4)
STAR = TokenType(3)
SLASH = TokenType(3)
PLUS = TokenType(2)
MINUS = TokenType(2)

NUMBER = TokenType(0)
LEFT_PAREN = TokenType(0)
RIGHT_PAREN = TokenType(0)


@dataclasses.dataclass()
class Token:
    type: TokenType
    literal: float | None
    start_at: int
    end_at: int


# It is a copy from "recursive_descent.py".
# ---
# But there are both the unary minus and the binary minus.
# It is important to distinguish them at this stage, because the shunting yard algorithm needs this info.
# ---
# And without EOF. I do not really understand the use of it.
class Scanner:

    def __init__(self, source):
        self._source = source
        self._tokens = []

        self._start = 0
        self._current = 0

    @property
    def is_at_end(self):
        return self._current >= len(self._source)

    def scan_tokens(self):
        while not self.is_at_end and self.peek() != '\n':
            self._start = self._current
            self.scan_token()

        return self._tokens

    def scan_token(self):
        c = self.advance()
        match c:
            case '+': self.add_token(PLUS)
            case '-':
                if (
                    not self._tokens or
                    (self._tokens[-1].type is not NUMBER and self._tokens[-1].type is not RIGHT_PAREN)
                ):
                    self.add_token(UMINUS)
                else:
                    self.add_token(MINUS)
            case '*': self.add_token(STAR)
            case '/': self.add_token(SLASH)
            case '(': self.add_token(LEFT_PAREN)
            case ')': self.add_token(RIGHT_PAREN)
            case ' ': pass
            case _:
                if self.is_digit(c):
                    self.number()
                else:
                    raise ParseError(f'Unexpected character: {c} at position {self._current-1}.')

    def advance(self):
        c = self._source[self._current]
        self._current += 1
        return c

    def add_token(self, type, literal=None):
        self._tokens.append(Token(type, literal, self._start, self._current-1))

    def is_digit(self, c):
        return 48 <= ord(c) <= 57

    def number(self):
        while self.is_digit(self.peek()):
            self.advance()

        if self.peek() == '.' and self.is_digit(self.peek_next()):
            self.advance()

            while self.is_digit(self.peek()):
                self.advance()

        self.add_token(NUMBER, float(self._source[self._start:self._current]))

    def peek(self):
        if self.is_at_end:
            return '\0'

        return self._source[self._current]

    def peek_next(self):
        if self._current + 1 >= len(self.source):
            return '\0'

        return self._source[self._current+1]

def parse(tokens):
    current = 0
    output_queue = []
    operators = []

    while current < len(tokens):
        token = tokens[current]
        current += 1

        if token.type is NUMBER:
            output_queue.append(token.literal)
            continue

        if token.type.precedence:
            while (
                operators
                and operators[-1].type is not LEFT_PAREN
                and operators[-1].type.precedence >= token.type.precedence
            ):
                output_queue.append(evaluate(operators.pop(), output_queue))

            operators.append(token)

        if token.type is LEFT_PAREN:
            operators.append(token)
            continue

        if token.type is RIGHT_PAREN:
            while operators and operators[-1].type is not LEFT_PAREN:
                output_queue.append(evaluate(operators.pop(), output_queue))

            if not operators:
                raise ParseError(f'Mismatched right paren at position {token.start_at}.')

            operators.pop()

    while operators:
        operator = operators.pop()
        if operator.type is LEFT_PAREN:
            raise ParseError(f'Mismatched left paren at position {operator.start_at}.')

        output_queue.append(evaluate(operator, output_queue))

    if len(output_queue) != 1:
        raise ParseError('Cannot evaluate the expression to the concrete value.')

    return output_queue.pop()

def evaluate(token, output_queue):
    operand = output_queue.pop()
    if token.type is PLUS:
        return output_queue.pop() + operand
    if token.type is MINUS:
        return output_queue.pop() - operand
    if token.type is UMINUS:
        return -operand
    if token.type is STAR:
        return output_queue.pop() * operand
    if token.type is SLASH:
        return output_queue.pop() / operand

    raise ParseError(f'Invalid token type (at pos {token.start_at}..{token.end_at}) when an operator expected.')

def get_source():
    with open('expression.txt', 'r') as f:
        return f.read()

if __name__ == '__main__':
    source = get_source()

    try:
        scanner = Scanner(source)
        tokens = scanner.scan_tokens()

        value = parse(tokens)
        print(f'Result: {value:.2f}')
    except ParseError as exc:
        print(exc, file=sys.stderr)
        sys.exit(65)

File: shunting_yard.rs
use std::fs;

#[derive(Debug)]
struct TokenType {
    uid: u8,
    precedence: u8,
}

impl PartialEq for TokenType {
    fn eq(&self, other: &Self) -> bool {
        self.uid == other.uid
    }
}

const UMINUS: TokenType = TokenType { uid: 1, precedence: 4 };
const STAR: TokenType = TokenType { uid: 2, precedence: 3 };
const SLASH: TokenType = TokenType { uid: 3, precedence: 3 };
const PLUS: TokenType = TokenType { uid: 4, precedence: 2 };
const MINUS: TokenType = TokenType { uid: 5, precedence: 2 };

const NUMBER: TokenType = TokenType { uid: 6, precedence: 0 };
const LEFT_PAREN: TokenType = TokenType { uid: 7, precedence: 0 };
const RIGHT_PAREN: TokenType = TokenType { uid: 8, precedence: 0 };

#[derive(Debug)]
struct Token {
    ttype: TokenType,
    lexeme: String,
    literal: Option<f64>,
    start_at: usize,
    end_at: usize,
}

struct Scanner {
    chars: Vec<char>,
    tokens: Vec<Token>,
    start: usize,
    current: usize,
}

impl Scanner {

    fn new(chars: Vec<char>) -> Self {
        Self {
            chars: chars,
            tokens: vec![],
            start: 0,
            current: 0,
        }
    }

    fn scan_tokens(&mut self) -> &Vec<Token> {
        while !self.is_at_end() && self.peek() != '\n' {
            self.start = self.current;
            self.scan_token();
        }

        return &self.tokens;
    }

    fn scan_token(&mut self) {
        let c = self.advance();
        match c {
            '+' => self.add_token(PLUS, None),
            '-' => {
                let tokens_count = self.tokens.len();
                if tokens_count < 1 || (
                    self.tokens[tokens_count-1].ttype != NUMBER &&
                    self.tokens[tokens_count-1].ttype != RIGHT_PAREN
                ) {
                    self.add_token(UMINUS, None);
                } else {
                    self.add_token(MINUS, None);
                }
            },
            '*' => self.add_token(STAR, None),
            '/' => self.add_token(SLASH, None),
            '(' => self.add_token(LEFT_PAREN, None),
            ')' => self.add_token(RIGHT_PAREN, None),
            _ => {
                if c == ' ' {
                    return;
                } if c.is_digit(10) {
                    self.number();
                } else {
                    panic!("Unexpected character: {} at position {}.", c, self.current-1);
                }
            },
        }
    }

    fn is_at_end(&self) -> bool {
        self.current >= self.chars.len()
    }

    fn advance(&mut self) -> char {
        let c = self.chars[self.current];
        self.current += 1;
        return c;
    }

    fn add_token(&mut self, ttype: TokenType, literal: Option<f64>) {
        self.tokens.push(Token {
            ttype: ttype,
            literal: literal,
            lexeme: self.make_lexeme(),
            start_at: self.start,
            end_at: self.current - 1,
        });
    }

    fn number(&mut self) {
        while self.peek().is_digit(10) {
            self.advance();
        }

        if self.peek() == '.' && self.peek_next().is_digit(10) {
            self.advance();

            while self.peek().is_digit(10) {
                self.advance();
            }
        }

        self.add_token(NUMBER, Some(self.get_float_number()));
    }

    fn peek(&self) -> char {
        if self.is_at_end() {
            return '\0';
        }

        return self.chars[self.current];
    }

    fn peek_next(&self) -> char {
        if self.current + 1 >= self.chars.len() {
            return '\0';
        }

        return self.chars[self.current+1]
    }

    // This way to make a substring is awful,
    // but I don't know how to do it in a better way.
    fn make_lexeme(&self) -> String {
        let mut string = String::new();
        for i in self.start..self.current {
            string.push(self.chars[i]);
        }
        return string;
    }

    fn get_float_number(&self) -> f64 {
        self.make_lexeme().parse::<f64>().unwrap()
    }

}

fn parse(tokens: &Vec<Token>) -> f64 {
    let mut current: usize = 0;
    let mut results: Vec<f64> = vec![];
    let mut operators: Vec<&Token> = vec![];

    let mut token: &Token;
    while current < tokens.len() {
        token = &tokens[current];
        current += 1;

        if token.ttype == NUMBER {
            let literal = match token.literal {
                Some(x) => x,
                None => unreachable!("Invalid number literal. Lexeme: {}.", token.lexeme),
            };
            results.push(literal);
        } else if token.ttype.precedence != 0 {
            while
                !operators.is_empty()
                && operators[operators.len()-1].ttype != LEFT_PAREN
                && operators[operators.len()-1].ttype.precedence >= token.ttype.precedence
            {
                let result = evaluate(operators.pop().unwrap(), &mut results);
                results.push(result);
            }
            operators.push(token);
        } else if token.ttype == LEFT_PAREN {
            operators.push(token);
        } else if token.ttype == RIGHT_PAREN {
            while !operators.is_empty() && operators[operators.len()-1].ttype != LEFT_PAREN {
                let result = evaluate(operators.pop().unwrap(), &mut results);
                results.push(result);
            }

            if operators.is_empty() {
                panic!("Mismatched right paren at position {}.", token.start_at);
            }

            operators.pop();
        }
    }

    while !operators.is_empty() {
        let operator = operators.pop().unwrap();
        if operator.ttype == LEFT_PAREN {
            panic!("Mismatched left paren at position {}.", operator.start_at);
        }

        let result = evaluate(operator, &mut results);
        results.push(result);
    }

    if results.len() != 1 {
        panic!("Cannot evaluate the expression to the concrete value.");
    }

    return results[0];
}

fn evaluate(operator: &Token, results: &mut Vec<f64>) -> f64 {
    let operand = results.pop().unwrap();

    if operator.ttype == PLUS {
        return results.pop().unwrap() + operand;
    }

    if operator.ttype == MINUS {
        return results.pop().unwrap() - operand;
    }

    if operator.ttype == UMINUS {
        return -operand;
    }

    if operator.ttype == STAR {
        return results.pop().unwrap() * operand;
    }

    if operator.ttype == SLASH {
        return results.pop().unwrap() / operand;
    }

    unreachable!("Invalid token type (at pos {}..{}) when an operator expected.", operator.start_at, operator.end_at);
}

fn get_source() -> String {
    match fs::read_to_string("expression.txt") {
        Err(why) => panic!("Couldn't read: {}", why),
        Ok(data) => data,
    }
}

fn main() {
    let source: String = get_source();

    let mut scanner = Scanner::new(source.chars().collect());
    let tokens = scanner.scan_tokens();

    // for token in tokens {
    //     println!("{:?}", token);
    // }

    println!("Result: {:.2}", parse(tokens));
}
